### PR TITLE
wasmtime 0.28.0 (new formula)

### PR DESCRIPTION
**Pull request by x13a**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/wasmtime.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class Wasmtime < Formula
+  desc "Standalone JIT-style runtime for WebAssembly, using Cranelift"
+  homepage "https://wasmtime.dev/"
+  license "Apache-2.0"
+  head "https://github.com/bytecodealliance/wasmtime.git"
+
+  stable do
+    url "https://github.com/bytecodealliance/wasmtime.git", tag: "v0.28.0"
+  end
```
Comment (carlocab) [inline Formula/wasmtime.rb:9]: This doesn't need to be a separate block. If we are using a Git clone, we also need to specify `revision:` in addition to `tag:`. Please see the other formulae that do this.

Also, why do we need a git clone? Can we use a tarball?

File: Formula/wasmtime.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  livecheck do
+    url :head
+    strategy :github_latest
```
Comment (carlocab) [inline Formula/wasmtime.rb:13]: Are you sure `:github_latest` is needed here? We usually try to avoid using it, as it entails making API calls that get rate-limited.

**Review by carlocab — COMMENTED** [review]
Thanks, @x13a. A few comments/questions.

File: Formula/wasmtime.rb (hunk 1 above)
Comment (x13a) [reply Formula/wasmtime.rb:9]: I suppose we can't use tarball because we need `git clone --recurse-submodules ..`. Building tarball won't work without it.

File: Formula/wasmtime.rb (hunk 2 above)
Comment (x13a) [reply Formula/wasmtime.rb:13]: There are some mess with tags, for example there is a tag `0.46.1` which is two years old and the latest version is `0.28.0`. That is why I used `github_latest` strategy here.

File: Formula/wasmtime.rb (hunk 1 above)
Comment (carlocab) [reply Formula/wasmtime.rb:9]: What is in the submodules?

File: Formula/wasmtime.rb (hunk 2 above)
Comment (carlocab) [reply Formula/wasmtime.rb:13]: Can we use a tighter regex to rule out the old tag instead? We can also use `:stable` instead of `:head` here.

File: Formula/wasmtime.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Wasmtime < Formula
+  desc "Standalone JIT-style runtime for WebAssembly, using Cranelift"
+  homepage "https://wasmtime.dev/"
+  url "https://github.com/bytecodealliance/wasmtime.git",
+      tag:      "v0.28.0",
+      revision: "e8b8947956a7fff6ce1475fafe6a13f15073570c"
+  license "Apache-2.0"
+  head "https://github.com/bytecodealliance/wasmtime.git"
```
Comment (carlocab) [inline Formula/wasmtime.rb:8]:
```suggestion
  head "https://github.com/bytecodealliance/wasmtime.git", branch: "main"
```

File: Formula/wasmtime.rb (hunk 1 above)
Comment (x13a) [reply Formula/wasmtime.rb]: Rust crates. The official build guide is:
```sh
git clone --recurse-submodules \
    https://github.com/bytecodealliance/wasmtime.git
cd wasmtime
cargo build --release
```

File: Formula/wasmtime.rb (hunk 2 above)
Comment (x13a) [reply Formula/wasmtime.rb:12]: Where are many tags, starting from `v0.31.0` and up to `v0.46.1`.

Comment (carlocab) [reply Formula/wasmtime.rb:12]: Ok, fair enough. CC @samford in case you have thoughts here.

Comment (samford) [reply Formula/wasmtime.rb:12]: The [`wasmtime` repository](https://github.com/bytecodealliance/wasmtime) was forked from the [now archived] [`cranelift` repository](https://github.com/bytecodealliance/cranelift). Unfortunately, this means that the `wasmtime` repo contains all the previous `cranelift` tags (e.g., `v0.46.1`). New `cranelift` tags in the `wasmtime` repo use a format like `cranelift-v0.69.0` but the older tags are indistinguishable from `wasmtime` versions.

As a result, we can't use the `Git` strategy with the standard regex for Git tags like `1.2.3`/`v1.2.3` (`v?(\d+(?:\.\d+)+)`), as the newest version will be incorrectly given as `0.46.1` (an old `cranelift` version) instead of `0.28.0` (the "latest" `wasmtime` version on GitHub). This would only work if upstream removed the old `cranelift` tags from the `wasmtime` repository.

Long story short, the `GithubLatest` strategy is unfortunately necessary here to identify the correct latest version, due to the weirdness in the upstream repository.

File: Formula/wasmtime.rb (hunk 1 above)
Comment (carlocab) [reply Formula/wasmtime.rb]: Ok, if they're just Rust crates that's fine. If it's something else though we'll need to have a closer look at it, as we don't want this to vendor things that are provided by Homebrew.

Comment (x13a) [reply Formula/wasmtime.rb]: This builds only one binary `wasmtime`. Nothing more.

Comment (carlocab) [reply Formula/wasmtime.rb]: It's not a question of what it builds. It's a matter of what the dependencies are, and there are some or else we wouldn't need submodules.

Comment (x13a) [reply Formula/wasmtime.rb]: From `git clone` output:
```text
Submodule 'crates/c-api/examples/wasm-c-api' (https://github.com/WebAssembly/wasm-c-api) registered for path 'crates/c-api/wasm-c-api'
Submodule 'WASI' (https://github.com/WebAssembly/WASI) registered for path 'crates/wasi-common/WASI'
Submodule 'crates/wasi-crypto/spec' (https://github.com/WebAssembly/wasi-crypto.git) registered for path 'crates/wasi-crypto/spec'
Submodule 'crates/wasi-nn/spec' (https://github.com/WebAssembly/wasi-nn) registered for path 'crates/wasi-nn/spec'
Submodule 'spec_testsuite' (https://github.com/WebAssembly/testsuite) registered for path 'tests/spec_testsuite'
```

**Review by carlocab — APPROVED** [review]
Thanks, @x13a.
